log4j-api/src/main/java/org/apache/logging/log4j/util/ServiceLoaderUtil.java

```diff
@@ -110,28 +78,73 @@ private ServiceLoaderUtil() {
      * @param logger      An action to perform for each broken service.
      * @param services    A list to add the services.
      */
-    private static <T> void addServices(final Class<T> serviceType, final ClassLoader classLoader,
-            final Consumer<Throwable> logger, final Collection<T> services) {
-        final Iterator<T> iterator = ServiceLoader.load(serviceType, classLoader).iterator();
-        while (iterator.hasNext()) {
-            try {
-                final T service = iterator.next();
-                if (classLoader.equals(service.getClass().getClassLoader())) {
-                    services.add(service);
-                }
-            } catch (ServiceConfigurationError e) {
-                if (logger != null) {
-                    logger.accept(e);
-                }
+    static <T> Stream<T> loadClassloaderServices(final Class<T> serviceType, final Lookup lookup,
+            final ClassLoader classLoader, final boolean verbose) {
+        return StreamSupport.stream(new ServiceLoaderSpliterator<T>(serviceType, lookup, classLoader, verbose), false);
+    }
+
+    static <T> Iterable<T> callServiceLoader(Lookup lookup, Class<T> serviceType, ClassLoader classLoader,
+            boolean verbose) {
+        try {
+            final MethodHandle handle = lookup.findStatic(ServiceLoader.class, "load", LOAD_CLASS_CLASSLOADER);
+            final ServiceLoader<T> serviceLoader = (ServiceLoader<T>) handle.invokeExact(serviceType, classLoader);
+            return serviceLoader;
+        } catch (Throwable e) {
+            if (verbose) {
+                StatusLogger.getLogger().error("Unable to load services for service {}", serviceType, e);
             }
         }
+        return Collections.emptyList();
+
     }
 
-    private static <T> void addOsgiServices(final Class<T> serviceType, final Collection<T> services) {
-        final Iterable<? extends T> iterable = org.glassfish.hk2.osgiresourcelocator.ServiceLoader
-                .lookupProviderInstances(serviceType);
-        if (iterable != null) {
-            iterable.forEach(services::add);
```

Comment:
This is not the right way to find OSGI services in Log4j: Log4j artifacts register OSGI services properly, so no need for the (probably) unsupported `osgi-resource-locator` or other tricks.